### DISCUSSION TITLE
Ideas for improving known security vulnerabilities in package ecosystems

### DISCUSSION THREAD

**Discussion by JonDouglas** [curl/curl#10657] [Ideas]
Hello,

I hope this discussion is the best place to provide some productive ideas for improving security vulnerabilities in package ecosystems that derive the curl functionality.

While I represent the NuGet group and saw a tweet here that sparked this discussion, this discussion mostly comes from my individual perspective. I may make mistakes, feel free to correct them. I'm trying to learn.

https://twitter.com/bagder/status/1630704007445372928

**The major challenge I see comes down to high quality information where a CNA can decipher affected software products for other ecosystems. And for that information to make it down to all the maintainers of said software.** 

Especially in this case where it is not practical nor curl's responsibility to list every official or unofficial ecosystem package as said affected software product.

<img width="662" alt="image" src="https://user-images.githubusercontent.com/1288848/222192435-e1d3876e-eb95-4413-9fe8-4c99675ae81a.png">

I'd like to inform a contemporary strategy I'm aware of that might help make this more manageable for the surrounding communities or even just create more confusion. Who knows.

I've seen other notable libraries like [TensorFlow](https://github.com/tensorflow/tensorflow/security/advisories) publish advisories [directly in their repository](https://docs.github.com/en/code-security/security-advisories/repository-security-advisories/about-repository-security-advisories). I noticed no such publication happens in this repository which might limit the visibility and potential for [community contributions](https://github.blog/2022-02-22-github-advisory-database-now-open-to-community-contributions/) of existing advisories. I did a little research and it looks like most are published on a [website here](https://curl.se/docs/security.html).

In practice, I could go to any of these known linked CVEs and community contribute an affected software package in an ecosystem I'm familiar with where it would be reviewed and be properly categorized which will bring security affordances to some toolchains. 

<img width="966" alt="image" src="https://user-images.githubusercontent.com/1288848/222196704-86c73d75-324c-4c0b-90af-4d9effec4a73.png">

The same could apply to curl:

<img width="1892" alt="image" src="https://user-images.githubusercontent.com/1288848/222201290-65c01388-3cdb-4069-b252-4d2d2b58e6c3.png">

This would really cover the case of not being able to contact a maintainer or not having a direct line with the existing CNA who issued the CVEs to update with the new information / issue new CVEs if the case warrants. In majority of cases, maintainers are quite responsive and just aren't aware of the vulnerability information making it downstream.

Anyway, just an idea as this is a hard problem to solve. This to me feels like a security vulnerability process inflection point. I don't think there's any right answers. I'm far from a security expert and I'm not fully aware of the end-to-end GitHub feature set either, but it has been helpful in my experiences.

**Comment by bagder** [comment]
This is a straight quote from https://github.com/github/advisory-database/ from just now

**Unfortunately, we cannot accept community contributions to advisories outside of our supported ecosystems.**

And then there's a list of "ecosystems" and none of them is an ecosystem where curl belongs. GitHub simply does not want our information.

We provide **ALL** info about **ALL** our published vulnerabilities at https://curl.se/docs/security.html. Not most, all. We also mention all vulnerabilities in release-notes and we email a separate dedicated email about each and every one to highlight them to users. The CVEs are also added and available [in this git repo](https://github.com/curl/curl-www/tree/master/docs). I don't believe they are hard to find for anyone who wants to learn about curl's past vulnerabilities.

**Comment by dfandrich** [comment]
The Open Source Vulnerability format https://ossf.github.io/osv-schema/ that
GitHub is using for their security reports is interesting. It's a
machine-readable (JSON) format including important bits like the first and last
affected versions so automated tools can determine whether a particular
project is affected.  It shouldn't be too hard to create OSV data from the curl
vulnerabilities at https://curl.se/docs/security.html. Much of the data are
already available in structured form.

But, if we were to do that, who would use it? There would have to be some
standard means of discovery for this to make a difference to more than a
handful of people. I note that https://osv.dev/ maintains a large database of
vulnerabilities in OSV format that tools can use, but I'm not sure how reports
are added. If they can check a curl RSS feed of advisories, then it's fairly
straightforward and could happen automatically. If not, then someone on the
curl team would need to remember to upload each new advisory, which seems prone
to failure.

**Reply by bagder** [reply]
Right, we could probably generate whatever format we want based on the data and metadata we already have without that being overly complicated. The question is what kind of infra that would use it or where we would provide it.

Possibly there is also a trust issue. We claim this is accurate data about curl vulns but will the world take our word for it?
